courses/tasks.py

```diff
@@ -0,0 +1,62 @@
+"""
+Tasks for the courses app
+"""
+import logging
+from mitxpro.celery import app
+from mitxpro.utils import now_in_utc
+from courses.models import CourseRun, CourseRunCertificate
+from courses.utils import ensure_course_run_grade, process_course_run_grade_certificate
+from courseware.api import get_edx_grades_with_users
+
+log = logging.getLogger(__name__)
+
+
+@app.task
+def generate_course_certificates():
+    """
+    Task to generate certificates for courses.
+    """
+    log.info("generate_course_certificates: Started")
+    now = now_in_utc()
+    course_runs = CourseRun.objects.filter(end_date__lt=now).exclude(
+        courseware_id__in=CourseRunCertificate.objects.values_list(
+            "course_run__courseware_id", flat=True
+        )
+    )
+
+    for run in course_runs:
+        edx_grade_user_iter = get_edx_grades_with_users(run)
+
+        for edx_grade, user in edx_grade_user_iter:
+            course_run_grade, created, updated = ensure_course_run_grade(
+                user=user, course_run=run, edx_grade=edx_grade
+            )
+
+            if created:
+                log.info("generate_course_certificates: New grade record created")
+            elif updated:
```

Comment:
I did not get you here, In my case, the flag --> `should_update` is `False` by default so I believe that line will not be executed ( But yes, need to look here why is it required then and possible changes. anyway ...  ) . 

So by reading your [that-comment ](https://github.com/mitodl/mitxpro/pull/1069/files#r320432060) , Do you mean that we should introduce a new Boolean field in `CourseRunGrade` model

`set_by_admin = models.BooleanField(default=False)` 
and the default value for `should_update` should be `True` [here](https://github.com/mitodl/mitxpro/blob/master/courses/utils.py#L7) but if `set_by_admin` is `True` in model, then `CourseRunGrade` should never be updated by edX grade data ?? Right ?

Reply:
@asadiqbal08 Here's what I'm thinking:

1. Add the `set_by_admin` to `CourseRunGrade` (and as you said, default it to `False`)
1. Set that flag to `True` only when the `sync_grades_and_certificates` command is run with a `grade` specified. Once it's set to `True`, it should stay set to `True` unless someone goes into Django admin and changes it.
1. Update `courses.utils.ensure_course_run_grade` logic:
  1. Under all circumstances, we should prevent a `CourseRunGrade` from being updated in this function if `set_by_admin=True`.
  1. We should keep the `should_update` flag just to give clients an option to ignore new data in the edX grades (but again, even if `should_update=True`, an update should not be performed if `set_by_admin=True`)
  1. As I mentioned in the comment at the start of this thread, it would be nice to avoid updating the `CourseRunGrade` record (and returning `True` for the `updated` flag) if that `CourseRunGrade` has exactly equivalent data to the edX grade.

As for changing the `should_update` default value, I don't have a very strong opinion on it so I'll let you decide. Let me know if any of that doesn't make sense